File: mobile/user-app/lib/features/channel/services/channel_service.dart
import '../../../auth.dart';
import '../models/channel_models.dart';

class ChannelService {
  ChannelService(this.auth);
  final AuthController auth;

  Future<ChannelDetail?> getMyChannel() async {
    try {
      final res = await auth.protected('GET', '/channels/me');
      return ChannelDetail.fromJson(res);
    } on ApiFailure catch (e) {
      if (e.status == 404) return null;
      rethrow;
    }
  }

  Future<ChannelDetail> getChannel(String handleOrId) async {
    final clean = handleOrId.startsWith('@')
        ? handleOrId.substring(1)
        : handleOrId;
    final res = await auth.protected(
      'GET',
      '/channels/handle/${Uri.encodeComponent(clean)}',
    );
    return ChannelDetail.fromJson(res);
  }

  Future<bool> checkHandle(String handle) async {
    final clean = handle.startsWith('@') ? handle.substring(1) : handle;
    try {
      final res = await auth.api.request(
        'GET',
        '/channels/check-handle?handle=${Uri.encodeQueryComponent(clean)}',
      );
      return res['isAvailable'] as bool? ?? false;
    } catch (_) {
      return false;
    }
  }

  Future<ChannelDetail> createChannel({
    required String name,
    required String handle,
    String? description,
  }) async {
    final clean = handle.startsWith('@') ? handle.substring(1) : handle;
    final res = await auth.protected(
      'POST',
      '/channels',
      body: {
        'name': name.trim(),
        'handle': clean.trim(),
        'description': description?.trim(),
      },
    );
    return ChannelDetail.fromJson(res);
  }

  Future<ChannelDetail> updateChannel(
    String id, {
    required String name,
    String? description,
  }) async {
    final res = await auth.protected(
      'PATCH',
      '/channels/${Uri.encodeComponent(id)}',
      body: {'name': name.trim(), 'description': description?.trim()},
    );
    return ChannelDetail.fromJson(res);
  }

  Future<void> deleteChannel(String id) async {
    await auth.protected('DELETE', '/channels/${Uri.encodeComponent(id)}');
  }

  Future<List<ChannelRole>> getRoles() async {
    final items = await auth.protectedList('/channels/roles');
    return items
        .whereType<Map<String, dynamic>>()
        .map(ChannelRole.fromJson)
        .toList();
  }

  Future<List<ChannelMember>> getMembers(String channelId) async {
    final items = await auth.protectedList('/channels/$channelId/members');
    return items
        .whereType<Map<String, dynamic>>()
        .map(ChannelMember.fromJson)
        .toList();
  }

  Future<List<ChannelInvitation>> getPendingInvitations(
    String channelId,
  ) async {
    final items = await auth.protectedList('/channels/$channelId/invitations');
    return items
        .whereType<Map<String, dynamic>>()
        .map(ChannelInvitation.fromJson)
        .toList();
  }

  Future<List<ChannelInvitation>> getMyInvitations() async {
    final items = await auth.protectedList('/channels/invitations/me');
    return items
        .whereType<Map<String, dynamic>>()
        .map(ChannelInvitation.fromJson)
        .toList();
  }

  Future<void> inviteMember(
    String channelId,
    String email,
    String roleCode,
  ) async {
    await auth.protected(
      'POST',
      '/channels/$channelId/members/invite',
      body: {'email': email.trim(), 'roleCode': roleCode},
    );
  }

  Future<void> acceptInvitation(String invitationId) async {
    await auth.protected('POST', '/channels/invitations/$invitationId/accept');
  }

  Future<void> declineInvitation(String invitationId) async {
    await auth.protected('POST', '/channels/invitations/$invitationId/decline');
  }

  Future<void> revokeInvitation(String channelId, String invitationId) async {
    await auth.protected(
      'DELETE',
      '/channels/$channelId/invitations/$invitationId',
    );
  }

  Future<void> changeMemberRole(
    String channelId,
    String userId,
    String roleCode,
  ) async {
    await auth.protected(
      'PATCH',
      '/channels/$channelId/members/$userId/role',
      body: {'roleCode': roleCode},
    );
  }

  Future<void> removeMember(String channelId, String userId) async {
    await auth.protected('DELETE', '/channels/$channelId/members/$userId');
  }
}

File: frontend/user-web/src/app/core/channel.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { RuntimeConfig } from './runtime-config';

export interface ChannelSummary {
  channelId: string;
  ownerUserId: string;
  name: string;
  handle: string;
  description: string | null;
  avatarUrl: string | null;
  bannerUrl: string | null;
  contactEmail: string | null;
  watermarkUrl: string | null;
  status: string;
  createdAt: string;
}

export interface ChannelDetail {
  channelId: string;
  ownerUserId: string;
  name: string;
  handle: string;
  description: string | null;
  avatarUrl: string | null;
  bannerUrl: string | null;
  contactEmail: string | null;
  watermarkUrl: string | null;
  settings: string;
  status: string;
  subscriberCount: number;
  videoCount: number;
  isOwner: boolean;
  myRole: ChannelRoleCode | null;
  permissions: string[];
  createdAt: string;
}

export type ChannelRoleCode = 'owner' | 'manager' | 'editor' | 'moderator' | 'viewer';

export interface ChannelRole {
  code: Exclude<ChannelRoleCode, 'owner'>;
  name: string;
  description: string;
  permissions: string[];
}

export interface ChannelMember {
  channelMemberId: string;
  channelId: string;
  userId: string;
  username: string;
  email: string;
  displayName: string;
  avatarUrl: string | null;
  roleCode: ChannelRoleCode;
  permissions: string[];
  status: string;
  joinedAt: string;
}

export interface ChannelInvitation {
  channelInvitationId: string;
  channelId: string;
  channelName: string;
  channelHandle: string;
  invitedUserId: string | null;
  invitedEmail: string | null;
  invitedByUserId: string;
  roleCode: Exclude<ChannelRoleCode, 'owner'>;
  permissions: string[];
  status: 'pending' | 'accepted' | 'declined' | 'expired' | 'revoked';
  expiresAt: string;
  createdAt: string;
}

export interface CreateChannelRequest {
  name: string;
  handle: string;
  description?: string;
  avatarUrl?: string;
  bannerUrl?: string;
  contactEmail?: string;
}

export interface UpdateChannelRequest {
  name?: string;
  handle?: string;
  description?: string;
  avatarUrl?: string;
  bannerUrl?: string;
  contactEmail?: string;
  watermarkUrl?: string;
  settings?: string;
}

export interface CheckHandleResponse {
  handle: string;
  isAvailable: boolean;
  message: string;
}

@Injectable({ providedIn: 'root' })
export class ChannelService {
  private http = inject(HttpClient);
  private config = inject(RuntimeConfig);
  private base = this.config.apiBaseUrl + '/channels';

  getMyChannel(): Observable<ChannelDetail> {
    return this.http.get<ChannelDetail>(`${this.base}/me`);
  }

  getAccessibleChannels(): Observable<ChannelDetail[]> {
    return this.http.get<ChannelDetail[]>(`${this.base}/accessible`);
  }

  getChannel(handle: string): Observable<ChannelDetail> {
    const clean = handle.startsWith('@') ? handle.slice(1) : handle;
    return this.http.get<ChannelDetail>(`${this.base}/handle/${encodeURIComponent(clean)}`);
  }

  createChannel(request: CreateChannelRequest): Observable<ChannelSummary> {
    return this.http.post<ChannelSummary>(this.base, request);
  }

  updateChannel(channelId: string, request: UpdateChannelRequest): Observable<ChannelSummary> {
    return this.http.patch<ChannelSummary>(`${this.base}/${channelId}`, request);
  }

  deleteChannel(channelId: string): Observable<void> {
    return this.http.delete<void>(`${this.base}/${channelId}`);
  }

  checkHandle(handle: string, currentChannelId?: string): Observable<CheckHandleResponse> {
    let params = new HttpParams().set('handle', handle);
    if (currentChannelId) {
      params = params.set('currentChannelId', currentChannelId);
    }
    return this.http.get<CheckHandleResponse>(`${this.base}/check-handle`, { params });
  }

  uploadAvatar(channelId: string, file: File): Observable<ChannelSummary> {
    const formData = new FormData();
    formData.append('file', file);
    return this.http.post<ChannelSummary>(`${this.base}/${channelId}/avatar`, formData);
  }

  uploadBanner(channelId: string, file: File): Observable<ChannelSummary> {
    const formData = new FormData();
    formData.append('file', file);
    return this.http.post<ChannelSummary>(`${this.base}/${channelId}/banner`, formData);
  }

  getRoles(): Observable<ChannelRole[]> {
    return this.http.get<ChannelRole[]>(`${this.base}/roles`);
  }

  getMembers(channelId: string): Observable<ChannelMember[]> {
    return this.http.get<ChannelMember[]>(`${this.base}/${channelId}/members`);
  }

  getPendingInvitations(channelId: string): Observable<ChannelInvitation[]> {
    return this.http.get<ChannelInvitation[]>(`${this.base}/${channelId}/invitations`);
  }

  getMyInvitations(): Observable<ChannelInvitation[]> {
    return this.http.get<ChannelInvitation[]>(`${this.base}/invitations/me`);
  }

  inviteMember(channelId: string, email: string, roleCode: ChannelRole['code']): Observable<ChannelInvitation> {
    return this.http.post<ChannelInvitation>(`${this.base}/${channelId}/members/invite`, { email, roleCode });
  }

  acceptInvitation(invitationId: string): Observable<ChannelMember> {
    return this.http.post<ChannelMember>(`${this.base}/invitations/${invitationId}/accept`, {});
  }

  declineInvitation(invitationId: string): Observable<void> {
    return this.http.post<void>(`${this.base}/invitations/${invitationId}/decline`, {});
  }

  revokeInvitation(channelId: string, invitationId: string): Observable<void> {
    return this.http.delete<void>(`${this.base}/${channelId}/invitations/${invitationId}`);
  }

  changeMemberRole(channelId: string, userId: string, roleCode: ChannelRole['code']): Observable<ChannelMember> {
    return this.http.patch<ChannelMember>(`${this.base}/${channelId}/members/${userId}/role`, { roleCode });
  }

  removeMember(channelId: string, userId: string): Observable<void> {
    return this.http.delete<void>(`${this.base}/${channelId}/members/${userId}`);
  }
}
